L1Trigger/DTTriggerPhase2/src/PseudoBayesGrouping.cc

```diff
@@ -0,0 +1,473 @@
+  //Compare L0-L6 and L1-L7
+  RecognisePatterns(digisinL0, digisinL6, L0L6Patterns);
+  RecognisePatterns(digisinL1, digisinL7, L1L7Patterns);
+  //Compare L0-L5, L1-L6, L2-L7
+  RecognisePatterns(digisinL0, digisinL5, L0L5Patterns);
+  RecognisePatterns(digisinL1, digisinL6, L1L6Patterns);
+  RecognisePatterns(digisinL2, digisinL7, L2L7Patterns);
+  //L0-L4, L1-L5, L2-L6, L3-L7
+  RecognisePatterns(digisinL0, digisinL4, L0L4Patterns);
+  RecognisePatterns(digisinL1, digisinL5, L1L5Patterns);
+  RecognisePatterns(digisinL2, digisinL6, L2L6Patterns);
+  RecognisePatterns(digisinL3, digisinL7, L3L7Patterns);
+  //L1-L4, L2-L5, L3-L6
+  RecognisePatterns(digisinL1, digisinL4, L1L4Patterns);
+  RecognisePatterns(digisinL2, digisinL5, L2L5Patterns);
+  RecognisePatterns(digisinL3, digisinL6, L3L6Patterns);
+  //L2-L4, L3-L5
+  RecognisePatterns(digisinL2, digisinL4, L2L4Patterns);
+  RecognisePatterns(digisinL3, digisinL5, L3L5Patterns);
+  //L3-L4
+  RecognisePatterns(digisinL3, digisinL4, L3L4Patterns);
+  //Uncorrelated SL1
+  RecognisePatterns(digisinL0, digisinL1, L0L1Patterns);
+  RecognisePatterns(digisinL0, digisinL2, L0L2Patterns);
+  RecognisePatterns(digisinL0, digisinL3, L0L3Patterns);
+  RecognisePatterns(digisinL1, digisinL2, L1L2Patterns);
+  RecognisePatterns(digisinL1, digisinL3, L1L3Patterns);
+  RecognisePatterns(digisinL2, digisinL3, L2L3Patterns);
+  //Uncorrelated SL3
+  RecognisePatterns(digisinL4, digisinL5, L4L5Patterns);
+  RecognisePatterns(digisinL4, digisinL6, L4L6Patterns);
+  RecognisePatterns(digisinL4, digisinL7, L4L7Patterns);
+  RecognisePatterns(digisinL5, digisinL6, L5L6Patterns);
+  RecognisePatterns(digisinL5, digisinL7, L5L7Patterns);
+  RecognisePatterns(digisinL6, digisinL7, L6L7Patterns);
+}
+
+void PseudoBayesGrouping::RecognisePatterns(std::vector<DTPrimitive> digisinLDown,
+                                            std::vector<DTPrimitive> digisinLUp,
+                                            std::vector<DTPattern*> patterns) {
+  //Loop over all hits and search for matching patterns (there will be four
+  // amongst ~60, accounting for possible lateralities)
+  for (auto dtPD_it = digisinLDown.begin(); dtPD_it != digisinLDown.end(); dtPD_it++) {
+    int LDown = dtPD_it->layerId();
+    int wireDown = dtPD_it->channelId();
+    for (auto dtPU_it = digisinLUp.begin(); dtPU_it != digisinLUp.end(); dtPU_it++) {
+      int LUp = dtPU_it->layerId();
+      int wireUp = dtPU_it->channelId();
+      int diff = wireUp - wireDown;
+      for (auto pat_it = patterns.begin(); pat_it != patterns.end(); pat_it++) {
+        //For each pair of hits in the layers search for the seeded patterns
+        if ((*pat_it)->sl1() != (LDown) || (*pat_it)->sl2() != (LUp) || (*pat_it)->diff() != diff)
+          continue;
+        //If we are here a pattern was found and we can start comparing
+        (*pat_it)->setHitDown(wireDown);
+        cand = new CandidateGroup(*pat_it);
```

Comment:
better to use a shared_ptr, also this probably doesn't have to be a member variable